bitsail-connectors/connector-ftp/src/main/java/com/bytedance/bitsail/connector/ftp/source/split/coordinator/FtpSourceSplitCoordinator.java

```diff
@@ -0,0 +1,176 @@
+
+    int readerNum = context.totalParallelism();
+    LOG.info("Found {} readers and {} splits.", readerNum, splitList.size());
+    if (readerNum > splitList.size()) {
+      LOG.error("Reader number {} is larger than split number {}.", readerNum, splitList.size());
+    }
+
+    for (FtpSourceSplit split : splitList) {
+      int readerIndex = ReaderSelector.getReaderIndex(readerNum);
+      splitAssignmentPlan.computeIfAbsent(readerIndex, k -> new HashSet<>()).add(split);
+      LOG.info("Will assign split {} to the {}-th reader", split.uniqSplitId(), readerIndex);
+    }
+  }
+
+  @Override
+  public void addReader(int subtaskId) {
+    LOG.info("Found reader {}", subtaskId);
+    tryAssignSplitsToReader();
+  }
+
+  @Override
+  public void addSplitsBack(List<FtpSourceSplit> splits, int subtaskId) {
+    LOG.info("Source reader {} return splits {}.", subtaskId, splits);
+
+    int readerNum = context.totalParallelism();
+    for (FtpSourceSplit split : splits) {
+      int readerIndex = ReaderSelector.getReaderIndex(readerNum);
+      splitAssignmentPlan.computeIfAbsent(readerIndex, k -> new HashSet<>()).add(split);
+      LOG.info("Re-assign split {} to the {}-th reader.", split.uniqSplitId(), readerIndex);
+    }
+
+    tryAssignSplitsToReader();
+  }
+
+  private void tryAssignSplitsToReader() {
+    Map<Integer, List<FtpSourceSplit>> splitsToAssign = new HashMap<>();
+
+    for (Integer readerIndex : splitAssignmentPlan.keySet()) {
+      if (CollectionUtils.isNotEmpty(splitAssignmentPlan.get(readerIndex)) && context.registeredReaders().contains(readerIndex)) {
+        splitsToAssign.put(readerIndex, Lists.newArrayList(splitAssignmentPlan.get(readerIndex)));
+      }
+    }
+
+    for (Integer readerIndex : splitsToAssign.keySet()) {
+      LOG.info("Try assigning splits reader {}, splits are: [{}]", readerIndex,
+              splitsToAssign.get(readerIndex).stream().map(FtpSourceSplit::getSplitId).collect(Collectors.toList()));
+      splitAssignmentPlan.remove(readerIndex);
+      context.assignSplit(readerIndex, splitsToAssign.get(readerIndex));
+      context.signalNoMoreSplits(readerIndex);
+      LOG.info("Finish assigning splits reader {}", readerIndex);
+    }
+  }
+
+  private List<FtpSourceSplit> construct(IFtpHandler ftpHandler) throws IOException {
+    List<FtpSourceSplit> splits = new ArrayList<>();
+    ftpHandler.loginFtpServer();
+    String[] paths = ftpHandler.getFtpConfig().getPaths();
+    int index = 0;
+    for (String path : paths) {
+      if (null != path && path.length() > 0) {
+        path = path.replace("\n", "").replace("\r", "");
+        String[] pathArray = StringUtils.split(path, ",");
+        for (String p : pathArray) {
+          for (String f : ftpHandler.getFiles(p.trim())) {
+            FtpSourceSplit split = new FtpSourceSplit(index++);
+            split.setPath(f);
+            splits.add(split);
+          }
+        }
+      }
+      this.fileSize += ftpHandler.getFilesSize(path);
```

Comment:
this info looks useful, maybe we can add this to the split

Reply:
Got it.  i add it to the FtpSourceSplit.